src/librbd/ManagedLock.cc

```diff
@@ -0,0 +1,672 @@
+  case STATE_LOCKED:
+  case STATE_REACQUIRING:
+  case STATE_PRE_SHUTTING_DOWN:
+  case STATE_POST_ACQUIRING:
+  case STATE_PRE_RELEASING:
+    lock_owner = true;
+    break;
+  default:
+    lock_owner = false;
+    break;
+  }
+
+  ldout(m_cct, 20) << "=" << lock_owner << dendl;
+  return lock_owner;
+}
+
+template <typename I>
+void ManagedLock<I>::shut_down(Context *on_shut_down) {
+  ldout(m_cct, 10) << dendl;
+
+  Mutex::Locker locker(m_lock);
+  assert(!is_shutdown_locked());
+  execute_action(ACTION_SHUT_DOWN, on_shut_down);
+}
+
+template <typename I>
+void ManagedLock<I>::acquire_lock(Context *on_acquired) {
+  int r = 0;
+  {
+    Mutex::Locker locker(m_lock);
+    if (is_shutdown_locked()) {
+      r = -ESHUTDOWN;
+    } else if (m_state != STATE_LOCKED || !m_actions_contexts.empty()) {
+      ldout(m_cct, 10) << dendl;
+      execute_action(ACTION_ACQUIRE_LOCK, on_acquired);
+      return;
+    }
+  }
+
+  on_acquired->complete(r);
+}
+
+template <typename I>
+void ManagedLock<I>::try_acquire_lock(Context *on_acquired) {
+  int r = 0;
+  {
+    Mutex::Locker locker(m_lock);
+    if (is_shutdown_locked()) {
+      r = -ESHUTDOWN;
+    } else if (m_state != STATE_LOCKED || !m_actions_contexts.empty()) {
+      ldout(m_cct, 10) << dendl;
+      execute_action(ACTION_TRY_LOCK, on_acquired);
+      return;
+    }
+  }
+
+  on_acquired->complete(r);
+}
+
+template <typename I>
+void ManagedLock<I>::release_lock(Context *on_released) {
+  int r = 0;
+  {
+    Mutex::Locker locker(m_lock);
+    if (is_shutdown_locked()) {
+      r = -ESHUTDOWN;
+    } else if (m_state != STATE_UNLOCKED || !m_actions_contexts.empty()) {
+      ldout(m_cct, 10) << dendl;
+      execute_action(ACTION_RELEASE_LOCK, on_released);
+      return;
+    }
+  }
+
+  on_released->complete(r);
+}
+
+template <typename I>
+void ManagedLock<I>::reacquire_lock(Context *on_reacquired) {
+  {
+    Mutex::Locker locker(m_lock);
+
+    if (m_state == STATE_WAITING_FOR_REGISTER) {
+      // restart the acquire lock process now that watch is valid
+      ldout(m_cct, 10) << ": " << "woke up waiting acquire" << dendl;
+      Action active_action = get_active_action();
+      assert(active_action == ACTION_TRY_LOCK ||
+             active_action == ACTION_ACQUIRE_LOCK);
+      execute_next_action();
+    } else if (!is_shutdown_locked() &&
+        (m_state == STATE_LOCKED || m_state == STATE_ACQUIRING)) {
+      // interlock the lock operation with other state ops
+      ldout(m_cct, 10) << dendl;
+      execute_action(ACTION_REACQUIRE_LOCK, on_reacquired);
+      return;
+    }
+  }
+
+  // ignore request if shutdown or not in a locked-related state
+  if (on_reacquired != nullptr) {
+    on_reacquired->complete(0);
+  }
+}
+
+template <typename I>
+void ManagedLock<I>::shutdown_handler(int r, Context *on_finish) {
+  on_finish->complete(r);
+}
+
+template <typename I>
+void ManagedLock<I>::pre_acquire_lock_handler(Context *on_finish) {
+  m_work_queue->queue(on_finish);
```

Comment:
I would feel better if it were possible to invoke all the ```xyz_handler``` methods without holding ```m_lock```. It would make for a consistent calling convention that the handlers are always invoked asynchronously and can invoke the provided ```on_finish``` w/o having to worry about lock ordering / recursive locking.